### PR TITLE
[WIP] Fea/scale w Async Queue

### PR DESCRIPTION
**Pull request by quasiben**
resolves https://github.com/dask/dask-kubernetes/issues/123

This PR currently implements a timeout with assertion that pods are in a `Running` state.  This is executed 3 times in try loop when scaling up clusters.  It's a very naive implementation and the timeout is hardcoded (this could an optional setting).  

We can improve this check with exponential backoff (this is how k8s works when asking for a resource) or implement a state-machine or internal queueing system to more closely evaluate state.  

More commits to come

### CODE REVIEW

File: dask_kubernetes/core.py [hunk 1]
```diff
@@ -508,6 +508,35 @@ def f(to_close):
     def scale_up(self, n, **kwargs):
         self.scheduler.loop.add_callback(self._scale_up, n, **kwargs)
 
+    async def _pod_status(self, pods):
```
Comment (mrocklin) [inline dask_kubernetes/core.py]:
```suggestion
   def _pod_status(self, pods):
```

This doesn't appear to need to be asynchronous

Comment (mrocklin) [reply dask_kubernetes/core.py]: Actually, this function only appears to be used once, and is relatively simple.  I recommend that we just inline this logic into the list comprehension where it is used.

File: dask_kubernetes/core.py [hunk 2]
```diff
@@ -508,6 +508,35 @@ def f(to_close):
...
+        for p in pods:
+            status[p.metadata.name] = p.status.phase
+        return status
+
+    async def check_status(self):
+        while True:
+            pods = await self._pods()
+            p_status = await self._pod_status(pods)
+            result = ['Running' == status for name, status in p_status.items()]
+            await asyncio.sleep(3.2)
```
Comment (mrocklin) [inline dask_kubernetes/core.py]: This seems like an odd place for a sleep, and also an odd value.  

I would expect to sleep only if we didn't pass everything, and I would also hope that the iteration time would be smaller than 3s.

File: dask_kubernetes/tests/test_async.py [hunk 3]
```diff
@@ -477,9 +477,7 @@ def pod_spec(image_name):
     assert set(cluster.scheduler.workers) == {b}
 
 
-@pytest.mark.xfail(
-    reason="The delay between scaling up, starting a worker, and then scale down causes issues"
-)
```
Comment (mrocklin) [inline dask_kubernetes/tests/test_async.py]: Can I ask you to also see if this resolves `test_scale_down_pending`?

File: dask_kubernetes/core.py [hunk 4]
```diff
@@ -508,6 +508,35 @@ def f(to_close):
...
+            if not all(result):
+                continue
```
Comment (mrocklin) [inline dask_kubernetes/core.py]: What if we scale down after scaling up within the timeout above?  Some of the pods might have a status of "done" or "cancelled" or something, right?

**Comment by mrocklin** [comment]
Another approach, which I'm not sure would scale well, would be for `scale` to store its most recent value on the Cluster object.

```python
>>> cluster.scale(20)
>>> cluster._manual_scale_target
20
```

And then we could periodically check the status of the pods, and make any changes if things had gotten out of alignment.  So if the last user request was for 20 pods and for some reason we only had 19, then we would ask Kubernetes for another.  If the user had asked for zero workers, and later we find out that surprisingly one has stuck around, then we would just try to delete it again.  This lets us be a bit sloppy about tracking individual requests, since we'll always try to bring things back in a dumb way to the user's defined goal.

At first I was concerned that this might get expensive with many pods (1000s is common enough) so it would be hard to do this check every second or so, but it looks like this all goes into one request, so maybe it isn't that bad?

Thoughts?

**Comment by quasiben** [comment]
I believe what you've outlined is similar to the state machine idea -- periodically check the request of and move towards matching request and status.  The periodic check can be tunable but I would expect at least a 5 second frequency?  When I've used/seen this used in the past it was in the context of a deployed thing being functional in <15 seconds.  Is that too slow ?  

Also, how do we check if a worker is stuck ?  Or do you mean the pod status changes from 'Running' to something else ?

While we discuss I'll build another PR which implements this idea

**Comment by quasiben** [comment]
In this PR I am going to assume that we want to handle adaptive scaling with low latencies.  This means being resilient to number of worker changes in a ~3 second window. 

- 0 s: 5 workers
- 3 s:  10 workers
- 6 s: 2 workers
...

In last commit (3d59f3d434f0fe75cd67a32aa31f687b7a0e3f8b), I've cheated a bit by checking the total number of workers vs the target -- abusing the suggestion of `_manual_scale_target`.

```python
result = [status_check == 'Running' for name, s in status.items()]
if sum(result) < self._manual_scale_target:
    continue
else:
    # we have met the target
    return True
````

By doing the above, we can be speedy about changes to the number of workers.  Updates to `_manual_scale_target` can be independent of our check of number of running workers.  Additionally, we are not bound by a periodic check -- the above is triggered when whenever we scale up.  As for scaling down, the code still simply calls delete and returns.  

The logic of scaling up is also simplified in that we try for 3 times to meet a target and timeout in each of those tries after `status-timeout` (default 5 seconds)

**Comment by quasiben** [comment]
fixing other tests now as well

File: dask_kubernetes/core.py [hunk 5]
```diff
@@ -500,27 +514,58 @@ def f(to_close):
...
 
-    async def _scale_up(self, n, pods=None, **kwargs):
+    async def check_status(self, status_check=PodStatus.running):
+        while True:
+            pods = await self._pods()
+            phases = {p.metadata.name: p.status.phase for p in pods}
+            result = [status_check.value == status for name, status in phases.items()]
+
+            # throttle loop
+            await asyncio.sleep(0.5)
```
Comment (mrocklin) [inline dask_kubernetes/core.py]: Ideally if things go smoothly the first time it would be good to not have any unnecessary slowdown.  This 500ms delay is probably better placed at the end of a loop like this, hopefully after a potential break/return.

File: dask_kubernetes/core.py [hunk 6]
```diff
@@ -614,7 +664,7 @@ def _cleanup_pods_sync(namespace, labels):
     for pod in pods.items:
         try:
             api.delete_namespaced_pod(
-                pod.metadata.name, namespace, kubernetes.client.V1DeleteOptions()
+                name=pod.metadata.name, namespace=namespace, body=kubernetes.client.V1DeleteOptions()
```
Comment (mrocklin) [inline dask_kubernetes/core.py]: +1 on explicit keywords, thanks

File: dask_kubernetes/core.py [hunk 7]
```diff
@@ -500,27 +514,58 @@ def f(to_close):
...
+            if result is None:
+                continue
+            elif sum(result) < self._manual_scale_target:
+                continue
+            else:
+                return True
+
+    async def status_with_timeout(self, timeout):
+        t = asyncio.create_task(self.check_status())
+        done, pending = await asyncio.wait({t}, timeout=timeout)
```
Comment (mrocklin) [inline dask_kubernetes/core.py]: I'm curious, is there a nicer way to do this?   Does asyncio require the `create_task` function to be run before `wait`?  (I don't know as much about asyncio as I do about tornado)

File: dask_kubernetes/core.py (hunk 5 above)
Comment (mrocklin) [reply dask_kubernetes/core.py]: As a reference point, for interactive workloads (which dask-kubernetes is sometimes used for) we probably are about anything over 100ms, especially if they start to build on each other.

If we're just doing a periodic cleanup check then yeah, a few seconds or whatever is fine.  But when you hit compute you want to see workers start up more or less immediately.

Comment (mrocklin) [reply dask_kubernetes/core.py]: The comment above is just for general context, not for this particular sleep.

File: dask_kubernetes/core.py [hunk 8]
```diff
@@ -28,6 +29,16 @@
...
 
 
+class PodStatus(Enum):
+    pending = 'Pending'
+    running = 'Running'
+    succeeded = 'Succeeded'
+    failed = 'Failed'
+    unknown = 'Unknown'
+    completed = 'Completed'
+    crashloopbackoff = 'CrashLoopBackOff'
```
Comment (mrocklin) [inline dask_kubernetes/core.py]: I have mixed thoughts about specifying things like this.  I imagine it's a matter of style.  I've personally always preferred using raw values directly rather than providing an abstraction layer between them.  This enum approach might be better as people use things like type systems and nicer editors though.  Maybe it's time for my style to change.

Comment (quasiben) [reply dask_kubernetes/core.py]: After chatting with @mrocklin we came to the conclusion that for small codebases and readability we should switch to global variables rather than introduce ENUMs

File: dask_kubernetes/core.py [hunk 9]
```diff
@@ -453,46 +484,53 @@ def logs(self, pod=None):
...
+                )
+                yield offload(self.scale_down, to_close)
+
+            self.scheduler.loop.add_callback(f, to_close)
+            return
 
-            # Terminate all pods without waiting for clean worker shutdown
-            self.scale_down(to_close)
+        # Terminate all pods without waiting for clean worker shutdown
+        return await self._scale_down(to_close)
```
Comment (mrocklin) [inline dask_kubernetes/core.py]: It seems like we want to wait for things to retire gracefully if we can, and then only shut things down hard if we weren't able to shut things down cleanly in a reasonable amount of time.

I would have expected us to just `await self.scheduler.retire_workers(...)` above rather than schedule it as a separate callback.

File: dask_kubernetes/core.py [hunk 10]
```diff
@@ -511,44 +549,18 @@ def f(to_close):
             except kubernetes.client.rest.ApiException as e:
                 # If a pod has already been removed, just ignore the error
                 if e.status != 404:
-                    raise
+                    raise kubernetes.client.rest.ApiException(e)
```
Comment (mrocklin) [inline dask_kubernetes/core.py]: Why this change?  Calling `raise` will raise the exception that we're curently handling.  I think it also retains the original traceback.

File: dask_kubernetes/core.py [hunk 11]
```diff
@@ -511,44 +549,18 @@ def f(to_close):
...
-        t = asyncio.create_task(self.check_status())
-        done, pending = await asyncio.wait({t}, timeout=timeout)
-
-        # if there are any pending tasks we have timed out
-        if len(pending):
-            return False
-
-        assert all([d.result() for d in done])
-        return True
+        self.task_queue.put_nowait(self._scale_down_with_pending_check)
```
Comment (mrocklin) [inline dask_kubernetes/core.py]: You may have wanted `_scale_up` here

File: dask_kubernetes/tests/test_async.py [hunk 12]
```diff
@@ -519,24 +519,28 @@ def pod_spec(image_name):
...
-        pytest.skip("Require a small test kubernetes cluster (maxpod <= 50)")
+        # It's probably not reasonable to run this test against a larger clusters
+        # artificially restrict to 50 pods
+        max_pods = 50
+
     extra_pods = 5
     requested_pods = max_pods + extra_pods
-    cluster.scale(requested_pods)
-
+    cluster.scale(20)
```
Comment (mrocklin) [inline dask_kubernetes/tests/test_async.py]: Why this change?

File: dask_kubernetes/core.py (hunk 9 above)
Comment (mrocklin) [reply dask_kubernetes/core.py]: Or maybe I should just say that there is value to waiting for the workers to transfer data before killing them.  Otherwise during adaptive workloads we end up repeating a lot of computation as our cluster scales itself down and loses results.

File: dask_kubernetes/core.py [hunk 13]
```diff
@@ -275,11 +280,37 @@ def asynchronous(self):
...
+                # Get a scale_up/down out of the queue.
+                func = await self.task_queue.get()
+                # execute function
+                await func()
+                # cleanup messy pods
+                pods = await self._cleanup_terminated_pods(pods)
+
+                self.task_queue.task_done()
+
+            await asyncio.sleep(1)
```
Comment (mrocklin) [inline dask_kubernetes/core.py]: I wonder if we might wait here for either 1s or when there is a new element in the task queue, whichever is first.  That might help to improve resonsiveness.

File: dask_kubernetes/core.py [hunk 14]
```diff
@@ -500,27 +512,58 @@ def f(to_close):
                     raise
 
     async def _cleanup_terminated_pods(self, pods):
-        terminated_phases = {"Succeeded", "Failed"}
+        terminated_phases = {PodStatus.succeeded, PodStatus.failed}
```
[the remark this answers is not in the document]
Comment (quasiben) [reply dask_kubernetes/core.py]: switched to globals

File: dask_kubernetes/core.py (hunk 7 above)
Comment (quasiben) [reply dask_kubernetes/core.py]: switched to a queueing system

File: dask_kubernetes/core.py (hunk 11 above)
Comment (quasiben) [reply dask_kubernetes/core.py]: yup!

File: dask_kubernetes/tests/test_async.py (hunk 12 above)
Comment (quasiben) [reply dask_kubernetes/tests/test_async.py]: a leftover from debugging

**Comment by quasiben** [comment]
This PR is currently failing a lot of tests on CircleCI.  I'm somewhat unfamiliar with CircleCI and I don't see where the output of the failing tests should be located.  I'm looking here:
https://circleci.com/gh/dask/dask-kubernetes/278

**Comment by mrocklin** [comment]
It would normally follow there, but that test is hanging I think, and so circle-ci maybe just stops the process entirely before pytest has a change to pill details?

You might try including a `-s` to include debug information during the tests, or you might try removing that test.

File: dask_kubernetes/core.py [hunk 15]
```diff
@@ -156,7 +163,7 @@ def __init__(
         env=None,
         auth=ClusterAuth.DEFAULT,
         asynchronous=False,
-        **kwargs
+        **kwargs,
```
Comment (mrocklin) [inline dask_kubernetes/core.py]: This will kill Python 3.5 compatibility.  This is likely a black issue.  You might want to update to black master and make sure that the rest of your file is compatible with Python 3.5.  If that's the case then black won't put these in there (I ran into this today as well, and found it frustrating).

File: dask_kubernetes/core.py [hunk 16]
```diff
@@ -440,45 +500,57 @@ def logs(self, pod=None):
 
         See Also
         --------
-        KubeCluster.scale_up
+        KubeCluster.scale_upc
```
Comment (mrocklin) [inline dask_kubernetes/core.py]:
```suggestion
        KubeCluster.scale_up
```

File: dask_kubernetes/core.py [hunk 17]
```diff
@@ -440,45 +500,57 @@ def logs(self, pod=None):
...
+        running_workers = list(self.scheduler.workers.keys())
+        running_ips = set(urlparse(worker).hostname for worker in running_workers)
+        running_names = [
+            v.name for w, v in self.scheduler.workers.items() if v.has_what
+        ]
+
+        # Dask is Fast! A worker can register with the scheduler before k8s
+        # moves the pod from `PENDING` to `RUNNING` below we remove
+        # any workers in a `PENDING` state which have non-zero
+        # memory or currently processing a task
```
Comment (mrocklin) [inline dask_kubernetes/core.py]: @jacobtomlinson have you run into this before?

File: dask_kubernetes/core.py (hunk 15 above)
Comment (quasiben) [reply dask_kubernetes/core.py]: Thanks for the note!

**Comment by quasiben** [comment]
@mrocklin finally have passing tests -- note I commented out test_sync.py.  Should I remove this file entirely ?

**Comment by mrocklin** [comment]
> @mrocklin finally have passing tests -- note I commented out test_sync.py. Should I remove this file entirely ?

Unfortunately we still need these tests to work.  These test that this code works when run from a notebook or normal script where the user isn't writing async/await code.

File: dask_kubernetes/tests/test_sync.py [hunk 18]
```diff
@@ -243,34 +243,34 @@ def test_pod_from_dict(image_name, loop, ns):
...
+#                     "image": image_name,
+#                     "imagePullPolicy": "IfNotPresent",
+#                     "name": "worker",
+#                 }
+#             ]
+#         }
+#     }
+
+#     with KubeCluster.from_dict(spec, loop=loop, namespace=ns) as cluster:
+#         cluster.adapt()
```
Comment (quasiben) [inline dask_kubernetes/tests/test_sync.py]: @mrocklin What should `cluster.adapt` do here ?  No workers have been requested yet

Comment (mrocklin) [reply dask_kubernetes/tests/test_sync.py]: `cluster.adapt()` starts up a coroutine that watches the activity in the scheduler and, if anything shows up, starts calling scale as necessary.

So, to start, nothing at all would happen.  But later on after we submit some work we would expect the cluster to automatically scale up.

**Comment by quasiben** [comment]
closing in favor of #162
